src/data/polynomial/trinomial.lean

```diff
@@ -0,0 +1,211 @@
+/-
+Copyright (c) 2020 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+
+import data.polynomial.reverse2
+import analysis.complex.polynomial
+
+/-!
+# Bundled Trinomials
+
+In this file we define an bundled trinomial, which are easier to
+work with than a sum of monomials.
+
+## Main definitions
+
+- `polynomial.trinomial`
+- `trinomial.of_polynomial`: unbundle a trinomial
+- `polynomial.to_trinomial`: bundle a trinomial
+
+## Main results
+
+- `reverse'_irreducible_test'`: an irreducibility criterion for unit trinomials
+- `reverse'_irreducible_test''`: an irreducibility criterion for unit trinomials
+- `selmer_irreducible`: the polynomial `X ^ n - X - 1` is irreducible for all `n ≠ 1`
+
+-/
+
+namespace polynomial
+
+/-- bundled trinomial -/
+structure trinomial (R : Type*) [semiring R] :=
+(a b c : R) (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠ 0) (i j k : ℕ) (hij : i < j) (hjk : j < k)
```

Comment:
Playing devil's advocate - would this definition be as easy to work with?
```
structure n_omial (R : Type*) (n : ℕ) [semiring R] :=
(coeffs : fin n → R)
(powers : fin n → ℕ)
(h_coeffs : ∀ i, coeffs i ≠ 0)
(h_powers : (list.of_fn powers).sorted (<))
```
because it scales much better to working with "quadnomials" etc

Reply:
It would certainly be better than working with polynomials. I'll have to play around with this definition a bit to see whether it is nice to work with.

Reply:
I've defined `to_polynomial` and `of_polynomial` for `n_omial`, and proved that they are inverses. So far, things are working ok (and the amount of code is about the same). I'll continue to see how my proof ports over.

Reply:
@eric-wieser Here are my thoughts so far:
1) It seems possible to develop an API for that definition, although some parts like `n_omial.reverse'` would be rather painful.
2) For the purposes of proving irreducibility of `X^n - X -1`, `trinomial` is much better than `n_omial`. The point of `trinomial` is that you can quickly figure what the `nat_degree` is or what the `reverse'` is (or other similar things). For `n_omial`, this is not the case. For example, the `nat_degree` of an `n_omial` is `coeff (fin.mk n.pred (nat.pred_lt n_ne_zero)`, and the `reverse'` of an `n_omial` is also messy (the cleanest definition that I can think of involves passing to a list and invoking list.reverse).

Reply:
`n_omial`s are sometimes called ["fewnomial"](https://en.wikipedia.org/wiki/Fewnomial_theory); see also the chapter on "Bernstein's theorem and Fewnomials" in Sturmfels's book [Solving systems of polynomial equations](https://math.berkeley.edu/~bernd/cbms.pdf) (though he never defines "fewnomial" formally, he just talks about polynomials with at most `m` distinct monomials).

Reply:
Isn't `fin.mk n.pred (nat.pred_lt n_ne_zero` just `fin.last`?

Reply:
Ah, I see that it's not. Could you push the progress you made on that definition to a branch?

Reply:
There's a bunch of errors at the end, but here's what I have so far: https://github.com/leanprover-community/mathlib/compare/nomial

Feel free to play around with it.

Reply:
`p.support.order_emb_of_fin` can replace your definition of `kth_power`, and produces an `order_embedding` which already has lots of API.

Reply:
I've pushed to your branch with that change.

Reply:
It seems that this lemma is missing:
```lean
f : fin n ↪o ℕ
⊢ (finset.image ⇑f finset.univ).order_emb_of_fin _ = f
``